sycl/source/handler.cpp

```diff
@@ -1065,6 +1065,43 @@ void handler::extractArgsAndReqs() {
   }
 }
 
+void handler::extractArgsAndReqsFromLambda(
+    char *LambdaPtr, detail::kernel_param_desc_t (*ParamDescGetter)(int),
+    size_t NumKernelParams, bool IsESIMD) {
+  size_t IndexShift = 0;
+  impl->MArgs.reserve(MaxNumAdditionalArgs * NumKernelParams);
```

Comment:
Aside from this line, why do we have to move this to the sources?

Reply:
To me, there is no strict necessity, just desire to move functionality out of public headers and out of templates.

Reply:
Removing the templates I don't mind, but the more symbols we have in the library, the less freedom we have due to the ABI it imposes. I could maybe also see some benefit in having this in the header w.r.t. inlining.
That is not to say I am against moving more stuff to the library, but when we do I prefer we move as much as we can to minimize the ABI surface.
Note that this is not a strong objection to moving this, but mainly for us to consider the pros/cons of moving small segments from the header to the library.

Reply:
This patch isn't based on the latest `origin/sycl`. Top of trunk has this (note `constexpr`):

https://github.com/intel/llvm/blob/0a406c9199016e05fc5cdbbbe518ca3f919f10de/sycl/include/sycl/handler.hpp#L769-L774

the only other thing we can fold into this library api is the `clearArgs` call (which might be a good thing to do, but doesn't really affect our freedom regarding ABI changes).

@steffenlarsen , I agree with you in general, but don't see how that is applicable in this particular case.

Reply:
My point here is, it seems like all that is actually forcing us to move this to the library is an optimization for pre-emptive reservation. It means the function we now have in the ABI does a handful of different things. Keeping it in the header and adding another new function in the library that does as little as possible still means we have another symbol, but I would argue that the chance of its signature changing or part of the ABI causing problems decreases with that design.

Reply:
`processArg` is only used in this function (https://github.com/search?q=repo%3Aintel%2Fllvm+processArg+path%3A%2F%5Esycl%5C%2F%2F&type=code), with this move we can remove it from the ABI. I'd argue that their interfaces are equally prone to cause ABI issues (mostly `IsESIMD` parameter).

That said, if you insist on moving this out of `.cpp` to `.hpp`, I won't object.

Reply:
The removal of `processArg` from the ABI is a good point. Could we do that under preview as part of this patch maybe?

Reply:
Am I right that current plan is to have `extractArgsAndReqsFromLambda()` logic internal and to prevent export of `handler::processArg()` at ABI breaking moment?

Reply:
IIUC, yes. @steffenlarsen , can you confirm?

Reply:
Yes, that plan sounds good to me!

Reply:
@steffenlarsen , what do you think?

Reply:
Yes! I like that solution, then we can move it when we promote preview during the next ABI break. ⭐